lib/multitask.py

```diff
@@ -0,0 +1,455 @@
+    return mixture_weights
+
+  def get_output_layer(num_classes, prefix):
+    return tkf.layers.Dense(num_classes, activation="softmax",
+                            name="%s_out" % prefix)
+  def rename(inputs, name):
+    return tkf.layers.Lambda(lambda x: x, name=name)(inputs)
+
+  num_classes = [datasets[x]["num_classes"] for x in datasets]
+  ds_names = list(datasets.keys())
+  num_datasets = len(datasets)
+
+  inputs = [tkf.Input(name="%s_in" % x, shape=input_shape) for x in ds_names]
+  if use_shared_mixture_weights:
+    mixture_weights = get_mixture_weights("shared", inputs)
+    features = [feature_extractor([x, *mixture_weights]) for x in inputs]
+  else:
+    mix_weights = [get_mixture_weights(name, inputs[0]) for name in ds_names]
+    features = [feature_extractor([inputs[i], *mix_weights[i]])
+                for i in range(num_datasets)]
+  if share_logits:
+    out_layer = get_output_layer(num_classes[0], "shared")
+    outputs = [rename(out_layer(features[i]), "%s_out" % ds_names[i])
+               for i in range(num_datasets)]
+  else:
+    out_layers = [get_output_layer(num_classes[i], ds_names[i])
+                  for i in range(num_datasets)]
+    outputs = [out_layers[i](features[i]) for i in range(num_datasets)]
+
+  combined_model = tkf.Model(inputs, outputs, name="combined")
+  return combined_model
+
+def get_combined_model(
+    datasets_info, input_shape, num_layers=16, num_templates=4,
+    tensor_size=16, in_adapter="strided", out_adapter="isometric",
+    dropout=0, kernel_reg=0, shared=False, share_mixture=False,
+    share_logits=False):
+  """Returns the multitask training model.
+
+  Arguments:
+  datasets_info: dataset info dictionary.
+  input_shape: shape of the input image.
+  num_layers: number of residual blocks in the feature extractor.
+  num_templates: number of templates.
+  tensor_size: size of the resblock input tensor.
+  in_adapter: input adapter type.
+  out_adapter: output adapter type.
+  dropout: dropout.
+  kernel_reg: kernel regularization parameter.
+  shared: if True, the shared model will be created.
+  share_mixture: if True, the mixture weights will be shared across
+  domains.
+  share_logits: if True, the logits layer will be shared.
+  """
+  if not shared:
+    feature_extractor = cresnet.resnet(
+        input_shape=input_shape, num_layers=num_layers,
+        num_classes=10, tensor_size=tensor_size, in_adapter=in_adapter,
+        out_adapter=out_adapter, kernel_regularizer=kernel_reg,
+        dropout=dropout, with_head=False, name="feature_extractor")
+    combined_model = build_models(
+        feature_extractor=feature_extractor, datasets=datasets_info,
+        input_shape=input_shape)
+  else:
+    feature_extractor = sresnet.shared_resnet(
+        input_shape=input_shape, num_layers=num_layers,
+        num_templates=num_templates, num_classes=10, tensor_size=tensor_size,
+        in_adapter=in_adapter, out_adapter=out_adapter,
```

Comment:
Many of these arguments can be shared (across `resnet` and `shared_resnet`) via a dictionary (with `**args` in the call).